In [45]:
!pip install pymysql sqlalchemy --quiet

In [46]:
import pandas as pd
from sqlalchemy import create_engine

df = pd.read_csv('../data/cleaned/tech_layoffs_cleaned.csv')

engine = create_engine('mysql+pymysql://root:616009@127.0.0.1:3306/tech_layoffs_analysis')

df.to_sql('layoffs', engine, if_exists='replace', index=False)

print("Import successful!")

Import successful!


**Business Question 1a: Which industries have experienced the most severe
layoffs by total headcount?**

In [47]:
query = """
SELECT Industry, SUM(Laid_Off) AS total_laid_off
FROM layoffs
GROUP BY Industry
ORDER BY total_laid_off DESC
"""
pd.read_sql(query, engine)

,Industry,total_laid_off
0,Other,126422.0
1,Consumer,65642.0
2,Retail,62190.0
3,Hardware,60842.0
4,Software Development,43634.0
5,Transportation,42980.0
6,Finance,42857.0
7,Food,39006.0
8,E-commerce,32434.0
9,IT Services and IT Consulting,22066.0


- Consumer, Retail, and Hardware top the list (excluding "Other," which 
aggregates low-frequency categories — see cleaning note), suggesting 
layoffs are concentrated in consumer-facing and hardware-dependent sectors. 
Software Development ranks 4th (43,634 layoffs) — significant, but notably 
below Consumer/Retail/Hardware, despite the dominant public narrative that 
software/engineering roles are the primary target of AI-driven layoffs.

**Business Question 1b: Which specific companies have had the most 
frequent layoff events?**

(Limited to companies with sufficient occurrences to avoid drawing 
conclusions from single, potentially anomalous events — see Company 
long-tail note from the exploration phase.)

In [48]:
query = """
SELECT Company, COUNT(*) AS layoff_times
FROM layoffs
GROUP BY Company
ORDER BY layoff_times DESC
LIMIT 15
"""
pd.read_sql(query, engine)

,Company,layoff_times
0,Google,16
1,Amazon,16
2,Microsoft,13
3,Intel,10
4,Meta,8
5,Chegg,7
6,eBay,7
7,Redfin,7
8,Rivian,7
9,Unity,6


**Comparing Question 1a and 1b: Frequency vs. Magnitude**

Industries like Consumer, Retail, and Hardware show high total layoff 
volume but relatively few companies driving that volume. In contrast, 
big tech companies (Google, Amazon: 16 events each; Microsoft: 13) show 
the opposite pattern — frequent, smaller-scale layoffs rather than one 
large event.

- This suggests two distinct corporate behaviors: large public tech 
companies appear to make incremental workforce adjustments repeatedly 
over time, while companies in Consumer/Retail/Hardware are more likely 
to execute fewer but larger-scale layoffs. This pattern is explored 
further in the next section (seasonal analysis).

In [49]:
query = """
SELECT Company, Date_layoffs, Laid_off
FROM layoffs
WHERE Company IN ('Amazon', 'Google', 'Microsoft')
ORDER BY Company
"""
pd.read_sql(query, engine)

,Company,Date_layoffs,Laid_off
0,Amazon,2022-10-28,150.0
1,Amazon,2022-11-16,10000.0
2,Amazon,2023-01-04,8000.0
3,Amazon,2025-05-24,100.0
4,Amazon,2025-07-17,NaN
5,Amazon,2023-03-20,9000.0
6,Amazon,2024-01-10,NaN
7,Amazon,2025-08-04,110.0
8,Amazon,2025-10-28,30000.0
9,Amazon,2024-01-18,30.0


In [50]:
query = """
SELECT MONTH(Date_layoffs) AS month_times,
	SUM(Laid_off) AS total_layoffs
FROM layoffs
GROUP BY MONTH(Date_layoffs)
ORDER BY total_layoffs DESC
"""
pd.read_sql(query, engine)

,month_times,total_layoffs
0,1,103839.0
1,4,100600.0
2,11,100339.0
3,5,64114.0
4,2,61059.0
5,3,57740.0
6,7,55108.0
7,8,54637.0
8,10,52943.0
9,6,45034.0


- Layoffs peak in January (103,839), April (100,600), and November (100,339), 
while December sees the fewest (19,162).

The January and April peaks align with corporate earnings cycles — January 
follows Q4 and annual earnings reports, and April follows Q1 earnings reports, 
suggesting layoffs are often executed as part of post-earnings workforce 
adjustments rather than random, isolated events.

The December low likely reflects companies avoiding layoffs during the 
holiday season for public image and employee morale reasons, with decisions 
potentially delayed to January instead.

**Implication for job seekers:** Layoff risk is not evenly distributed 
throughout the year. Job seekers evaluating an offer that starts in 
January, April, or November may want to factor in this seasonal pattern — 
not as a reason to avoid these months, but as context for understanding 
why a company might announce layoffs shortly after they join.

**Business Question 3: Is there a relationship between company size and 
layoff severity?**

Using the size categories established during data cleaning (Small ≤300, 
Mid-size 301-1000, Large >1000), this compares the average layoff 
percentage across company sizes — does company size correlate with how 
severely a company cuts its workforce when layoffs happen?

In [51]:
query = """
SELECT Company_Size_Category, 
	ROUND(AVG(Percentage), 2) AS layoff_percentage
FROM layoffs
GROUP BY Company_Size_Category
"""
pd.read_sql(query, engine)

,Company_Size_Category,layoff_percentage
0,Small,35.50
1,Mid-size,19.38
2,Large,11.97
3,None,58.84


**Note on NULL category:** Records with `Company_Size_Category = NULL` 
are companies where `Company_Size_before_Layoffs` was missing in the 
original dataset — this is expected behavior, not an error, since the 
binning function cannot categorize missing values.

Interestingly, this NULL group shows the highest average layoff percentage 
(58.84%), which may further support the "smaller companies cut deeper" 
pattern — companies with undocumented size are plausibly smaller, 
less-established firms. However, this is a secondary observation and 
should be verified against sample size before drawing strong conclusions.

**Key Finding:**

Company size shows a clear, consistent relationship with layoff severity: 
Small companies cut an average of 35.5% of their workforce per layoff event, 
compared to 19.38% for Mid-size and just 11.97% for Large companies.

This supports the earlier Stage-based finding from the exploration phase — 
small, less-established companies tend to make deeper, more existential cuts, 
while large companies make smaller, more incremental adjustments (consistent 
with the "frequent, small-scale" pattern observed for Google, Amazon, and 
Microsoft in Question 1b).

**Implication for job seekers:** Company size is a meaningful risk signal — 
not just whether a company has had layoffs, but how severe those layoffs 
tend to be, varies substantially by company size.

**Business Question 4: Does funding stage correlate with layoff severity?**

Comparing average layoff percentage across funding stages (Seed, Series 
A-J, Post-IPO, Acquired, etc.) to see whether earlier-stage companies show 
different layoff patterns than mature, later-stage companies.

In [52]:
query = """
SELECT STAGE, 
	ROUND(AVG(Percentage), 2) AS layoff_percentage
FROM layoffs
GROUP BY Stage
ORDER BY layoff_percentage DESC
"""
pd.read_sql(query, engine)

,STAGE,layoff_percentage
0,Seed,72.50
1,Series A,41.08
2,None,34.51
3,Series B,33.51
4,Acquired,31.93
5,Subsidiary,26.56
6,Series C,23.01
7,Series D,19.60
8,Series G,16.81
9,Series E,16.81


Layoff percentage decreases as funding stage matures: Seed-stage companies 
average 72.5% workforce reduction per layoff event, compared to single 
digits for Series H/I (8.78%-10.98%).

⚠️ **Important caveat:** This pattern is partially explained by a base-rate 
effect — early-stage companies have much smaller total headcounts, so even 
a small number of layoffs translates into a high percentage. A company 
with 5 employees laying off 2 people shows 40% — but this doesn't 
necessarily indicate more "severe" or "irresponsible" cuts than a large 
company laying off 500 out of 100,000 (0.5%). 

This finding should be read alongside Question 3 (company size), which 
shows the same directional pattern, but both metrics share this same 
base-rate limitation. A more rigorous follow-up analysis would examine 
absolute headcount changes alongside percentage, not percentage alone.

**Business Question 5: Does geographic location influence layoff patterns?**

Using `Country` (instead of `Region`, which was dropped due to being too 
coarse — see cleaning notes), comparing total layoffs and average layoff 
percentage across countries. Note: USA accounts for 64.6% of records, 
likely reflecting both genuine industry concentration and data collection 
bias toward English-language sources (see BRD Section 5).

In [53]:
query = """
SELECT Country, 
       COUNT(*) AS layoff_events, 
       SUM(Laid_Off) AS total_layoffs
FROM layoffs
GROUP BY Country
ORDER BY layoff_events DESC
LIMIT 10
"""
pd.read_sql(query, engine)

,Country,layoff_events,total_layoffs
0,USA,1557,541854.0
1,India,177,55212.0
2,Israel,122,14000.0
3,Canada,110,14195.0
4,Germany,81,28933.0
5,UK,79,17518.0
6,Brazil,48,7029.0
7,Australia,34,3164.0
8,Singapore,28,5316.0
9,Sweden,24,17201.0


In [54]:
query = """
SELECT Country, 
       COUNT(*) AS layoff_events, 
       SUM(Laid_Off) AS total_layoffs,
       ROUND(SUM(Laid_Off) / COUNT(*), 0) AS avg_per_event
FROM layoffs
GROUP BY Country
ORDER BY layoff_events DESC
LIMIT 10
"""
pd.read_sql(query, engine)

,Country,layoff_events,total_layoffs,avg_per_event
0,USA,1557,541854.0,348.0
1,India,177,55212.0,312.0
2,Israel,122,14000.0,115.0
3,Canada,110,14195.0,129.0
4,Germany,81,28933.0,357.0
5,UK,79,17518.0,222.0
6,Brazil,48,7029.0,146.0
7,Australia,34,3164.0,93.0
8,Singapore,28,5316.0,190.0
9,Sweden,24,17201.0,717.0


**New Question: Why is Sweden's average layoff size (717) so 
much higher than other countries?**

With only 24 layoff events, Sweden's average is highly sensitive to 
outliers — similar to the right-skewed distribution issue identified 
earlier with Money_Raised_in__mil. Before accepting this as a genuine 
country-level pattern, checking whether this average is driven by one 
or two unusually large layoff events.

In [ ]:
query = """
SELECT Company, Laid_Off, Date_layoffs
FROM layoffs
WHERE Country = 'Sweden'
ORDER BY Laid_Off DESC;
"""
pd.read_sql(query, engine)

,Company,Laid_Off,Date_layoffs
0,Ericsson,8500.0,2023-02-24
1,Northvolt,2800.0,2025-03-31
2,Northvolt,1600.0,2024-09-23
3,Spotify,1500.0,2023-12-04
4,Klarna,700.0,2022-05-23
5,Spotify,600.0,2023-01-23
6,Kry,300.0,2022-10-31
7,Spotify,200.0,2023-06-05
8,Voi,130.0,2022-12-07
9,Voi,120.0,2024-02-16


: 

**Investigation result: Sweden's high average is driven by a single outlier**

Ericsson's single layoff event (8,500 employees, Feb 2023) accounts for 
the majority of Sweden's total layoffs — roughly equal to the combined 
total of all other 23 events. This confirms the average of 717/event is 
not representative of a broader "Swedish layoff pattern"; it is a 
statistical artifact of one company's outsized event.

**Implication:** Country-level averages should not be interpreted at face 
value when sample sizes are small (as with Sweden's 24 events). This 
reinforces the broader lesson from this analysis: aggregated metrics 
(means, percentages) can mask the underlying distribution, and a single 
large event can distort conclusions — especially for countries or 
categories with fewer data points.

**Am unexpected interesting find: What actually happened at Ericsson?**

Ericsson's 8,500-employee layoff (Feb 2023) was driven by slowing 5G 
equipment demand (particularly in North America) and a cost-cutting target 
of $880 million by end of 2023, following weaker-than-expected Q4 earnings. 
This was a global cut affecting ~8% of Ericsson's workforce; only 1,400 of 
those positions were in Sweden, which is the portion captured in this dataset.

Notably, this case differs from the broader "over-hiring correction" 
narrative established earlier (Question 2) — Ericsson had not significantly 
expanded headcount in 2021, and this layoff actually brought its workforce 
to its lowest level since 2010. This suggests telecom-hardware companies 
may face layoff drivers distinct from the software/internet sector 
(declining product demand and margin pressure, rather than correcting 
pandemic-era overhiring).

**Takeaway:** Not all layoffs in this dataset share the same root cause. 
The "three-phase" narrative (pandemic shock → over-hiring correction → 
AI restructuring) is a useful general pattern, but individual company 
cases — like Ericsson — show that industry-specific dynamics also matter, 
and a single dominant narrative shouldn't be applied uniformly across 
all companies.

*Source: CNBC, Reuters, TechHQ (Feb 2023 reporting)*